Lib/typing.py

```diff
@@ -1850,6 +1850,11 @@ def __instancecheck__(cls, instance):
         if super().__instancecheck__(instance):
             return True
 
+        # quick true negative check for protocols with many attributes.
+        if not hasattr(instance, '__getattr__'):
+            if not (cls.__protocol_attrs__ <= set(dir(instance))):
```

Comment:
What about classes that define custom `__dir__()` methods? What about classes that define custom `__getattribute__` methods? This approach is very risky

Reply:
Well, is there any C-implemented alternative to `dir`, that returns some iterable of the attribute names? I saw `inspect.getmembers_static`, but it is way too slow.

Reply:
dir itself is even inefficient since it sorts the names, here we really only need something that iterates over the names.

Reply:
In regard to the custom `__getattribute__` stuff, I think the current code already suffers from that, no? The docs say about `getattr_static`:

> Note: this function may not be able to retrieve all attributes that getattr can fetch (like dynamically created attributes) and may find attributes that getattr can’t (like descriptors that raise AttributeError).

Reply:
> is there any C-implemented alternative to `dir`, that returns some iterable of the attribute names?

Not that I know of

Reply:
> In regard to the custom `__getattribute__` stuff, I think the current code already suffers from that, no?

Ah indeed; good point

Reply:
So really the issue here seems that

1. `getattr_static` is just really slow (I re-ran the test, replacing it with plain `getattr` and got 11x speedup)
   - `getattr` can be slow if it triggers some computation.
2. `isinstance(x, proto)` only checks statically known attributes and can give incorrect results if certain attributes are defined dynamically.
3. There is nothing in the data-model for classes to provide something like a `__hasattr__` to communicate which attributes are set dynamically.

Reply:
Your analysis is correct.

> * `getattr_static` is just really slow (I re-ran the test, replacing it with plain `getattr` and got 11x speedup)
>   
>   * `getattr` can be slow if it triggers some computation.
> * `isinstance(x, proto)` only checks statically known attributes and can give incorrect results if certain attributes are defined dynamically.

Note that we changed these two bullet points in Python 3.12 after long discussions in https://github.com/python/typing/issues/1363 and https://github.com/python/cpython/issues/102433. Previously, `isinstance()` checks _didn't_ use `inspect.getattr_static`, and _did_ resolve dynamically defined attributes. But it was decided that that behaviour was bad, so we decided to accept the performance penalty in order to have more correct and less unexpected behaviour. (And through various other performance optimisations to compensate for that, `isinstance()` checks against runtime-checkable protocols are still overall much faster in Python 3.12 than in Python 3.11, despite the slowdown from switching to `inspect.getattr_static`.)